cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/SimpleProgramOptions.java

```diff
@@ -64,4 +64,16 @@ public Arguments getUserArguments() {
   public boolean isDebug() {
     return debug;
   }
+
+  @Override
+  public long getLogicalStartTime() {
+    // use logical start time from arguments if present
+    // if not, use logical start time from user arguments if present
+    // if not, use current system time
+    String logicalStartTimeStr = arguments.hasOption(ProgramOptionConstants.LOGICAL_START_TIME) ?
+      arguments.getOption(ProgramOptionConstants.LOGICAL_START_TIME) :
+      userArguments.getOption(ProgramOptionConstants.LOGICAL_START_TIME);
+    return logicalStartTimeStr == null ?
+      System.currentTimeMillis() : Long.parseLong(logicalStartTimeStr);
```

Comment:
I think we should determine it in the constructor instead of on demand so that we make sure when current time millis is used, it is close to the actual start time of the program. The first time it gets created is in the `ProgramLifecycleService.start`. All subsequence instantiation is basically copy on write operation, which mean we should preserve the logical start time.